palette/moreland/luminance.go

```diff
@@ -0,0 +1,256 @@
+func (l *luminance) At(scalar float64) (color.Color, error) {
+	if l.max == 0 {
+		return nil, fmt.Errorf("moreland: Luminance color map Max == 0")
+	}
+	scalar = scalar / l.max
+	if scalar < 0 || scalar > 1 {
+		return nil, fmt.Errorf("moreland: interpolation value (%g) out of range (0,%g)", scalar, l.max)
+	}
+	i := sort.SearchFloat64s(l.scalars, scalar)
+	if i == 0 {
+		return l.colors[i].cieXYZ().rgb().sRGBA(l.alpha), nil
+	}
+	c1 := l.colors[i-1]
+	c2 := l.colors[i]
+	frac := (scalar - l.scalars[i-1]) / (l.scalars[i] - l.scalars[i-1])
+	o := cieLAB{
+		L: frac*(c2.L-c1.L) + c1.L,
+		A: frac*(c2.A-c1.A) + c1.A,
+		B: frac*(c2.B-c1.B) + c1.B,
+	}.cieXYZ().rgb().sRGBA(l.alpha)
+	o.clamp()
+	return o, nil
+}
+
+// SetMax implements the palette.ColorMap interface for a luminance value.
+func (l *luminance) SetMax(v float64) {
+	l.max = v
+}
+
+// SetMin implements the palette.ColorMap interface for a luminance value.
+// However, it will panic whenever it is called because the
+// minimum value must always be zero.
+func (l *luminance) SetMin(v float64) {
+	panic("moreland: Luminance minimum value cannot be changed from zero")
+}
+
+// Max implements the palette.ColorMap interface for a luminance value.
+func (l *luminance) Max() float64 {
+	return l.max
+}
+
+// Min implements the palette.ColorMap interface for a luminance value.
+func (l *luminance) Min() float64 {
+	return 0
+}
+
+// SetAlpha sets the opacity value of this color map. Zero is transparent
+// and one is completely opaque.
+// The function will panic is alpha is not between zero and one.
+func (l *luminance) SetAlpha(alpha float64) {
+	if alpha > 1 || alpha < 0 {
+		panic(fmt.Errorf("invalid alpha: %g", alpha))
+	}
+	l.alpha = alpha
+}
+
+// Alpha returns the opacity value of this color map.
+func (l *luminance) Alpha() float64 {
+	return l.alpha
+}
+
+// Palette returns a value that fulfills the palette.Palette interface,
+// where nColors is the number of desired colors.
+func (l luminance) Palette(n int) palette.Palette {
+	if l.max == 0 {
+		l.max = 1
+	}
+	delta := l.max / float64(n-1)
+	var v float64
+	c := make([]color.Color, n)
+	for i := 0; i < n; i++ {
+		var err error
+		c[i], err = l.At(v)
+		if err != nil {
+			panic(err)
+		}
+		v += delta
+	}
+	return plte(c)
+}
+
+// plte fulfils the palette.Palette interface.
+type plte []color.Color
+
+// Colors fulfils the palette.Palette interface.
+func (p plte) Colors() []color.Color {
+	return p
+}
+
+// BlackBody is a Luminance-class ColorMap based on the colors of black body radiation.
+// Although the colors are inspired by the wavelengths of light from
+// black body radiation, the actual colors used are designed to be
+// perceptually uniform. Colors of the desired brightness and hue are chosen,
+// and then the colors are adjusted such that the luminance is perceptually
+// linear (according to the CIE LAB color space).
+func BlackBody() palette.ColorMap {
```

Comment:
The struct has fields that need to be mutable (max, alpha, etc), so we need to create a new copy for each use because there may be multiple instances of BlackBody that have different max values, etc.